### PR TITLE
Improve time series data splitting

### LINKED ISSUE #3287: Improve Time Series Evaluation Algorithm
For this discussion, I'm using the zillow san francisco price dataset here:

### The problem - perform worse than baseline
[zillow-housing-prices-train-fixed-sfbay-daily.csv.zip](https://github.com/alteryx/evalml/files/7946371/zillow-housing-prices-train-fixed-sfbay-daily.csv.zip)

If you run this dataset through AutoMLSearch right now with the following parameters,

```
{"gap": 0,
 "forecast_horizon": 5,
 "max_delay": 11,
  "time_index": "ts"}
```

you'll see that the time series baseline performs the best:

image

### Investigation
To get a better idea of what was happening, I plotted the actual data vs the prediction for each cross validation fold for the elastic net regressor:

image

The top row is the first fold and so on. Two things jumped out at me:
1. The performance is not too bad on the first and third fold but the performance on the second fold is bad 
image
2. The predictions stretch out for 60 rows (5 years)! I don't think anyone in the world can predict housing prices for five years out with absolute frequency.

Here is a close up of the first fold:
image

I think 1 and 2 are related. The reason why the model performs so poorly on the second fold is because it's trained on data where housing prices rise and then asked to predict on the following *5 years* of data where the market goes down, and it's never seen the market go down! 

We shouldn't be evaluating the model on how well it predicts *5 years* of data when the user asks us to find a model that can predict *5 months* (forecast_horizon=5).

### Takeaway
We should update out AutoMLSearch evaluation algorithm to evaluate the pipelines on `forecast_horizon` number of rows in each cross validation fold. 

### Proof of concept
As a proof of concept, I modified the AutoMLSearch algorithm to evaluate each pipeline on `forecast_horizon` number of rows in each cross validation fold. The predictions more closely match the actual data! The reason is because we let the model iteratively adapt to changes in the underlying distribution

image

The pipelines now perform better than the baseline

image

### PR DESCRIPTION
**Pull request by eccabay**
Closes #3287

### CODE REVIEW

**Review by chukarsten — APPROVED** [review]
OK, just some minor nits.  I think I get how this is working...it seems like a super minor change made, passing the forecast horizon into the TSTimeSplit had major results.  Do we need to update the docs at all?

File: evalml/tests/conftest.py [hunk 1]
```diff
@@ -1757,6 +1757,12 @@ def test_context(
             return_value=predict_proba_return_value,
             pipeline_class_str=pipeline_to_mock,
         )
+        mock_predict = self._patch_method(
```
Comment (jeremyliweishih) [inline evalml/tests/conftest.py:1760]: think this is leftover from the draft 😄

File: evalml/tests/automl_tests/test_time_series_split.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ def test_time_series_split_init():
 
 @pytest.mark.parametrize(
     "gap,max_delay,forecast_horizon,n_splits",
-    [[7, 3, 1, 4], [0, 3, 2, 3], [1, 1, 1, 4]],
+    [[7, 3, 1, 5], [0, 8, 2, 3], [5, 4, 2, 4]],
```
[the remark this answers is not in the document]
Comment (eccabay) [reply evalml/tests/automl_tests/test_time_series_split.py:26]: Should I refactor this to make it clearerer?

File: evalml/utils/gen_utils.py [hunk 3]
```diff
@@ -609,18 +609,19 @@ def are_ts_parameters_valid_for_split(
...
+    train_size = n_obs - eval_size
     window_size = gap + max_delay + forecast_horizon
     msg = ""
-    if split_size <= window_size:
+    if train_size <= window_size:
         msg = (
             f"Since the data has {n_obs} observations and n_splits={n_splits}, "
-            f"the smallest split would have {split_size} observations. "
-            f"Since {gap + max_delay + forecast_horizon} (gap + max_delay + forecast_horizon) >= {split_size}, "
+            f"the smallest split would have {train_size} observations. "
```
Comment (jeremyliweishih) [inline evalml/utils/gen_utils.py:619]: Isn't `train_size` influenced by `forecast_horizon`? Should we update line 618 to mention `forecast_horizon` as well?

**Review by jeremyliweishih — APPROVED** [review]
Looks good to me - some suggestions for improvements and some clarification questions. @eccabay and I spoke offline about the changes to the forecast horizon in the tests with the root cause being having a `forecast_horizon` of 1 errors out on scoring for `AUC` since it requires more than 1 class. The proposed solution is to remove `AUC` as a default objective for TS classification and @eccabay will file an issue to track.

File: evalml/preprocessing/data_splitters/time_series_split.py [hunk 4]
```diff
@@ -56,16 +56,18 @@ def __init__(
...
     ):
         self.max_delay = max_delay
         self.gap = gap
-        self.forecast_horizon = forecast_horizon
+        self.forecast_horizon = forecast_horizon if forecast_horizon else 1
         self.time_index = time_index
         self.n_splits = n_splits
-        self._splitter = SkTimeSeriesSplit(n_splits=n_splits)
+        self._splitter = SkTimeSeriesSplit(
+            n_splits=n_splits, test_size=forecast_horizon or None
```
[the remark this answers is not in the document]
Comment (eccabay) [reply evalml/preprocessing/data_splitters/time_series_split.py:69]: Yeah, this was an oversight on my part. There was an issue with the forecast horizon defaulting to 1 caused issues so this was a workaround, but it's not actually necessary after updating the default value. I'll fix this.

File: evalml/preprocessing/data_splitters/time_series_split.py [hunk 5]
```diff
@@ -56,16 +56,18 @@ def __init__(
...
         gap=0,
-        forecast_horizon=1,
+        forecast_horizon=None,
         time_index=None,
         n_splits=3,
     ):
         self.max_delay = max_delay
         self.gap = gap
-        self.forecast_horizon = forecast_horizon
+        self.forecast_horizon = forecast_horizon if forecast_horizon else 1
```
[the remark this answers is not in the document]
Comment (eccabay) [reply evalml/preprocessing/data_splitters/time_series_split.py:65]: This situation is a little tricky, but I think we should keep `forecast_horizon` as an optional parameter. 

Previously, the test size was None for scikit's `TimeSeriesSplit`, so it defaulted to `n_samples // (n_splits + 1)`. The only reason `forecast_horizon` was used in this function was to validate the split size was large enough. 

Notice here that `self.forecast_horizon` gets set to 1 if `forecast_horizon` is None. However, `forecast_horizon` and not `self.forecast_horizon` is what's passed in to the `SkTimeSeriesSplit`. This way, if no forecast horizon is passed in, we maintain our historical behavior and only update the split size when a forecast horizon is explicitly set. Does that make sense?

File: evalml/tests/automl_tests/test_automl.py [hunk 6]
```diff
@@ -4612,14 +4614,14 @@ def test_cv_validation_scores_time_series(
...
-        "forecast_horizon": 1,
+        "forecast_horizon": 2,
     }
     automl = AutoMLSearch(
         X_train=X,
         y_train=y,
         problem_type="time series binary",
         max_iterations=3,
-        data_splitter=TimeSeriesSplit(n_splits=3),
+        data_splitter=TimeSeriesSplit(n_splits=3, forecast_horizon=2),
```
[the remark this answers is not in the document]
Comment (eccabay) [reply evalml/tests/automl_tests/test_automl.py:4624]: Without the `forecast_horizon`, the `TimeSeriesSplit` would have larger test sizes in the cv splits. However, that doesn't actually affect this test in any way, so I'll back it out, along with any other unnecessary similar changes!

**Pull request merged**